Services/UICore/classes/class.ilTemplate.php

```diff
@@ -335,12 +335,20 @@ function get($part = "DEFAULT", $add_error_mess = false,
 	* Set message. Please use ilUtil::sendInfo(), ilUtil::sendSuccess()
 	* and ilUtil::sendFailure()
 	*/
-	function setMessage($a_type, $a_txt, $a_keep = false)
+	public function setMessage($a_type, $a_txt, $a_keep = false)
 	{
 		if (!in_array($a_type, array("info", "success", "failure", "question")) || $a_txt == "")
 		{
 			return;
 		}
+		global $DIC;
+		if ($DIC->http()
+		        ->request()
+		        ->getQueryParams()[ilFileStandardDropzoneInputGUI::ASYNC_FILEUPLOAD]
+		    && !$a_keep) {
+			echo json_encode([ 'success' => ($a_type !== 'failure'), 'message' => $a_txt ]);
+			exit;
+		}
```

Comment:
ilTemplate should not depend on ilFileStandardDropzoneInputGUI. Again it seems to be a general "json reponse" mode. ilCtrl could handle such a parameter for now as a workaround, since it currently is handling request/response related things. Maybe "responseMode=json" in the request would trigger this and its a constant  like ilCtrl::RESPONSE_JSON. I am pretty sure that you will have better ideas in the future. I guess this might all go to something that extends $DIC->http() in the future.

Both the form and the template class can depend on a general json mode being obtained from ilCtrl for now imo.